matrix_io/malos/drivers.py

```diff
@@ -0,0 +1,180 @@
+https://github.com/matrix-io/matrix-creator-malos
+
+List of base ports
+------------------
+
+IMU with port 20013.
+Humidity with port 20017.
+Everloop with port 20021.
+Pressure with port 20025.
+UV with port 20029.
+MicArray_Alsa with port 20037.
+
+Each port reserves a range of 4 ports that are used for a driver:
+
+* Base port: (driver port) port for sending configuration proto to MALOS driver.
+* Keep alive port: port for sending keep alive messages, so MALOS driver keeps reading data
+  from sensors.
+* Error port: port yielding MALOS errors when they occur.
+* Data update port: port yielding sensor data updates.
+
+"""
+
+import asyncio
+import zmq
+
+from zmq.asyncio import Context
+
+IMU_PORT = 20013
+HUMIDITY_PORT = 20017
+EVERLOOP_PORT = 20021
+PRESSURE_PORT = 20025
+UV_PORT = 20029
+MICARRAY_ALSA_PORT = 20037
+
+
+def driver_port(address, base_port, ctx, config_proto):
+    """
+    MALOS configuration port
+
+    Uses the base port and accepts a configuration proto, the driver.proto
+    to configure the driver.
+
+    Details about the proto structure can be found here:
+    https://github.com/matrix-io/protocol-buffers/blob/master/malos/driver.proto
+
+    Args:
+        address: IP address of the device exposing the MALOS 0MQ sockets
+        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
+        ctx: a 0MQ context
+        config_proto: a driver.proto containing configuration for the driver
+    """
+
+    # Set up socket as a push
+    sock = ctx.socket(zmq.PUSH)
+
+    # Connect to the config port, same base port
+    sock.connect('tcp://{0}:{1}'.format(address, base_port))
+
+    # Send the configuration
+    sock.send(config_proto.SerializeToString())
+    sock.close()
+
+
+async def keep_alive_port(address, base_port, ctx, delay=5.0):
+    """
+    MALOS keep-alive port
+
+    Connects to the corresponding keep-alive port (base_port +1) given the desired
+    base_port. The keep alive port will run as long as the data_port is yielding
+    messages from 0MQ.
+
+    Args:
+        address: IP address of the device exposing the MALOS 0MQ sockets
+        base_port: desired base port as in List of base ports above (ie. 20013, 20017)
+        ctx: a 0MQ context
+        delay: delay between pings
+    """
+
+    # Set up socket as a push
+    sock = ctx.socket(zmq.PUSH)
+
+    # Connect to the keep alive port to the sensor port from the function args + 1
+    sock.connect('tcp://{0}:{1}'.format(address, base_port + 1))
+
+    # If the keep-alive pings stop, MALOS will stop the driver and stop sending
+    # updates. Pings are useful to prevent blocked applications from keeping
+    # a MALOS driver bussy.
+    while True:
+        try:
+            # Ping with empty string to let the drive know we're still listening
+            await sock.send_string('')
+
+            # Delay between next ping
+            await asyncio.sleep(delay)
+        except asyncio.CancelledError:
```

Comment:
Does it make sense to print other exceptions? Even if you don't die on them.

Reply:
In this case any other exception will propagate. We are catching the `CancelledError`  as in this case we just want to exit the `while True`  loop gracefully. References here https://docs.python.org/3/library/asyncio-task.html

Reply:
However, when using ensure_future, exceptions will not crash the system and might go unnoticed, so I will try to work around this to make the code easier to debug.